File: leetcode-04/solutions/python3/0581.py
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        min_ = float('inf')
        max_ = float('-inf')
        flag = False

        for i in range(1, len(nums)):
            if nums[i] < nums[i - 1]:
                flag = True
            if flag:
                min_ = min(min_, nums[i])

        flag = False

        for i in range(len(nums) - 2, -1, -1):
            if nums[i] > nums[i + 1]:
                flag = True
            if flag:
                max_ = max(max_, nums[i])

        for l in range(len(nums)):
            if nums[l] > min_:
                break

        for r in range(len(nums) - 1, -1, -1):
            if nums[r] < max_:
                break

        return 0 if r - l <= 0 else r - l + 1

File: leetcode-04/solutions/cpp/0581.cpp
class Solution {
 public:
  int findUnsortedSubarray(vector<int>& nums) {
    int min = INT_MAX;
    int max = INT_MIN;
    bool flag = false;

    for (int i = 1; i < nums.size(); i++) {
      if (nums[i] < nums[i - 1]) flag = true;
      if (flag) min = std::min(min, nums[i]);
    }

    flag = false;

    for (int i = nums.size() - 2; i >= 0; i--) {
      if (nums[i] > nums[i + 1]) flag = true;
      if (flag) max = std::max(max, nums[i]);
    }

    int l;
    int r;

    for (l = 0; l < nums.size(); l++)
      if (nums[l] > min) break;

    for (r = nums.size() - 1; r >= 0; r--)
      if (nums[r] < max) break;

    return r - l <= 0 ? 0 : r - l + 1;
  }
};